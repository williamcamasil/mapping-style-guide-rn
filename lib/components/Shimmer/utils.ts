import { ViewStyle } from 'react-native';

type ShimmerBasePropsType = {
  margin?: number;
  marginVertical?: number;
  marginHorizontal?: number;
  marginStart?: number;
  marginEnd?: number;
  marginTop?: number;
  marginBottom?: number;
};

export type ShimmerListOptionsType = ShimmerBasePropsType & {
  contentList: Array<ShimmerListOptionsType | ShimmerItemOptionsType>;
  flexDirection?: ViewStyle['flexDirection'];
};

export type ShimmerItemOptionsType = ShimmerBasePropsType & {
  height: number;
  width: number;
};

export function isListShimmer(item: ShimmerListOptionsType | ShimmerItemOptionsType): item is ShimmerListOptionsType {
  return 'contentList' in item;
}
