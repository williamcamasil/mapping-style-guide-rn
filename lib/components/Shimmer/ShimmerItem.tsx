import React, { useMemo } from 'react';
import { StyleSheet } from 'react-native';

import { LinearGradient } from 'expo-linear-gradient';
import { createShimmerPlaceholder } from 'react-native-shimmer-placeholder';

import { useViewStyles } from '../../hooks';
import { AppThemeType } from '../../theme';
import { ShimmerItemOptionsType } from './utils';

type ShimmerItemPropsType = {
  item: ShimmerItemOptionsType;
  theme: AppThemeType;
};

const ShimmerPlaceHolder = createShimmerPlaceholder(LinearGradient);

const styles = StyleSheet.create({
  shimmer: {
    flexDirection: 'row',
  },
});

const ShimmerItem: React.FC<ShimmerItemPropsType> = ({ item, theme }) => {

  const shimmerColors = useMemo(
    () => [
      theme.colors.neutralGray300,
      theme.colors.neutralGray200,
      theme.colors.neutralGray300],
    [theme],
  );

  const shimmerStyle = useViewStyles(() => [
    styles.shimmer,
    { borderRadius: theme.borders.radius.medium },
    item,
  ], [theme, item]);

  return (
    <ShimmerPlaceHolder
      shimmerColors={shimmerColors}
      shimmerStyle={shimmerStyle}
    />
  );
};

export default ShimmerItem;
