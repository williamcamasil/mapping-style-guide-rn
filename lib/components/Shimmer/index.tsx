import React from 'react';
import {
  View,
} from 'react-native';

import { AppThemeType, withTheme } from '../../theme';
import ShimmerItem from './ShimmerItem';
import { isListShimmer, ShimmerItemOptionsType, ShimmerListOptionsType } from './utils';

export type {
  ShimmerItemOptionsType, ShimmerListOptionsType,
} from './utils';

type ShimmerPropsType = ShimmerListOptionsType & {
  theme: AppThemeType;
  visible?: boolean;
  testId?: string;
};

const Shimmer: React.FC<ShimmerPropsType> = ({
  theme, contentList, visible, testId, ...others
}) => {
  if (!visible) return null;

  const renderShimmerItem = (item: ShimmerListOptionsType | ShimmerItemOptionsType, index: number) => {
    if (isListShimmer(item)) {
      return (
        <Shimmer
          {...item}
          theme={theme}
          key={index}
          visible
        />
      );
    }

    return (
      <ShimmerItem item={item} key={index} theme={theme} />
    );
  };

  return (
    <View testID={testId} style={others}>
      {contentList.map(renderShimmerItem)}
    </View>
  );
};

export default withTheme(Shimmer);
