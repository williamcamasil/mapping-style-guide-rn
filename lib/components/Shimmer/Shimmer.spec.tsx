import React from 'react';

import { render } from '@testing-library/react-native';

import Shimmer from '.';

describe('Shimmer snapshots', () => {
  it('empty', () => {
    const tree = render(<Shimmer visible contentList={[]} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('simple shimmer', () => {
    const tree = render(<Shimmer
      visible
      contentList={[
        { height: 10, width: 10 },
        { height: 10, width: 10 },
        { height: 10, width: 10 },
      ]}
    />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('nested shimmer', () => {
    const tree = render(<Shimmer
      visible
      contentList={[
        {
          contentList: [
            { height: 10, width: 10 },
            { height: 10, width: 10 },
            { height: 10, width: 10 },
          ],
        },
      ]}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('invisible shimmer', () => {
    const tree = render(<Shimmer
      visible={false}
      contentList={[
        {
          contentList: [
            { height: 10, width: 10 },
            { height: 10, width: 10 },
            { height: 10, width: 10 },
          ],
        },
      ]}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
