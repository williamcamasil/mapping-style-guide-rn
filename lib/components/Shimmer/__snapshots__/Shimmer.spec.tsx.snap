// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shimmer snapshots empty 1`] = `
<View
  style={{}}
/>
`;

exports[`Shimmer snapshots invisible shimmer 1`] = `null`;

exports[`Shimmer snapshots nested shimmer 1`] = `
<View
  style={{}}
>
  <View
    style={{}}
  >
    <View
      style={
        [
          {
            "height": 15,
            "width": 200,
          },
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 12,
            "flexDirection": "row",
            "height": 10,
            "width": 10,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "height": 0,
              "opacity": 0,
              "width": 0,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "backgroundColor": "#E1E2E4",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "flex": 1,
              "transform": [
                {
                  "translateX": -200,
                },
              ],
            }
          }
        >
          <View
            colors={
              [
                "#E1E2E4",
                "#F0F0F2",
                "#E1E2E4",
              ]
            }
            end={
              {
                "x": 2,
                "y": 0.5,
              }
            }
            locations={
              [
                0.3,
                0.5,
                0.7,
              ]
            }
            start={
              {
                "x": -1,
                "y": 0.5,
              }
            }
            style={
              {
                "flex": 1,
                "width": 200,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "height": 15,
            "width": 200,
          },
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 12,
            "flexDirection": "row",
            "height": 10,
            "width": 10,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "height": 0,
              "opacity": 0,
              "width": 0,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "backgroundColor": "#E1E2E4",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "flex": 1,
              "transform": [
                {
                  "translateX": -200,
                },
              ],
            }
          }
        >
          <View
            colors={
              [
                "#E1E2E4",
                "#F0F0F2",
                "#E1E2E4",
              ]
            }
            end={
              {
                "x": 2,
                "y": 0.5,
              }
            }
            locations={
              [
                0.3,
                0.5,
                0.7,
              ]
            }
            start={
              {
                "x": -1,
                "y": 0.5,
              }
            }
            style={
              {
                "flex": 1,
                "width": 200,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "height": 15,
            "width": 200,
          },
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 12,
            "flexDirection": "row",
            "height": 10,
            "width": 10,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "height": 0,
              "opacity": 0,
              "width": 0,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "backgroundColor": "#E1E2E4",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "flex": 1,
              "transform": [
                {
                  "translateX": -200,
                },
              ],
            }
          }
        >
          <View
            colors={
              [
                "#E1E2E4",
                "#F0F0F2",
                "#E1E2E4",
              ]
            }
            end={
              {
                "x": 2,
                "y": 0.5,
              }
            }
            locations={
              [
                0.3,
                0.5,
                0.7,
              ]
            }
            start={
              {
                "x": -1,
                "y": 0.5,
              }
            }
            style={
              {
                "flex": 1,
                "width": 200,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Shimmer snapshots simple shimmer 1`] = `
<View
  style={{}}
>
  <View
    style={
      [
        {
          "height": 15,
          "width": 200,
        },
        {
          "overflow": "hidden",
        },
        {
          "borderRadius": 12,
          "flexDirection": "row",
          "height": 10,
          "width": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "height": 0,
            "opacity": 0,
            "width": 0,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "backgroundColor": "#E1E2E4",
          "flex": 1,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "transform": [
              {
                "translateX": -200,
              },
            ],
          }
        }
      >
        <View
          colors={
            [
              "#E1E2E4",
              "#F0F0F2",
              "#E1E2E4",
            ]
          }
          end={
            {
              "x": 2,
              "y": 0.5,
            }
          }
          locations={
            [
              0.3,
              0.5,
              0.7,
            ]
          }
          start={
            {
              "x": -1,
              "y": 0.5,
            }
          }
          style={
            {
              "flex": 1,
              "width": 200,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "height": 15,
          "width": 200,
        },
        {
          "overflow": "hidden",
        },
        {
          "borderRadius": 12,
          "flexDirection": "row",
          "height": 10,
          "width": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "height": 0,
            "opacity": 0,
            "width": 0,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "backgroundColor": "#E1E2E4",
          "flex": 1,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "transform": [
              {
                "translateX": -200,
              },
            ],
          }
        }
      >
        <View
          colors={
            [
              "#E1E2E4",
              "#F0F0F2",
              "#E1E2E4",
            ]
          }
          end={
            {
              "x": 2,
              "y": 0.5,
            }
          }
          locations={
            [
              0.3,
              0.5,
              0.7,
            ]
          }
          start={
            {
              "x": -1,
              "y": 0.5,
            }
          }
          style={
            {
              "flex": 1,
              "width": 200,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "height": 15,
          "width": 200,
        },
        {
          "overflow": "hidden",
        },
        {
          "borderRadius": 12,
          "flexDirection": "row",
          "height": 10,
          "width": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "height": 0,
            "opacity": 0,
            "width": 0,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "backgroundColor": "#E1E2E4",
          "flex": 1,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "transform": [
              {
                "translateX": -200,
              },
            ],
          }
        }
      >
        <View
          colors={
            [
              "#E1E2E4",
              "#F0F0F2",
              "#E1E2E4",
            ]
          }
          end={
            {
              "x": 2,
              "y": 0.5,
            }
          }
          locations={
            [
              0.3,
              0.5,
              0.7,
            ]
          }
          start={
            {
              "x": -1,
              "y": 0.5,
            }
          }
          style={
            {
              "flex": 1,
              "width": 200,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
